astropy/cosmology/core.py

```diff
@@ -84,12 +84,96 @@ class CosmologyError(Exception):
+    name : str or None (optional, keyword-only)
+        The name of the cosmology.
+    **kwargs
+        Arguments into the cosmology; used by subclasses, not this base class.
+
+    Notes
+    -----
+    Class instances are static -- you cannot (and should not) change the values
+    of the parameters.  That is, all of the attributes above are
+    read only.
+
+    """
+
+    def __init_subclass__(cls):
+        super().__init_subclass__()
+        # get signature, dropping "self" by taking arguments [1:]
+        sig = signature(cls.__init__)
+        sig = sig.replace(parameters=list(sig.parameters.values())[1:])
+        cls._init_signature = sig  # store (immutable) initialization signature
+
+    def __new__(cls, *args, **kwargs):
+        # bundle initialization argument
+        if not args:
+            parameters = kwargs
+        else:  # have to merge args and kwargs.
+            # bind any arguments passed
+            ba = cls._init_signature.bind_partial(*args, **kwargs)
+            ba.apply_defaults()  # and fill in the defaults
+            # get dictionary of arguments
+            parameters = ba.arguments
+
+        self = super().__new__(cls)
+        self._init_arguments = parameters  # store arguments from initialization
+
+        return self
+
+    def __init__(self, *args, name=None, **kwargs):
+        self.name = name
+
+    def clone(self, **kwargs):
+        """Returns a copy of this object with updated parameters, as specified.
+
+        This cannot be used to change the type of the cosmology, so ``clone()``
+        cannot be used to change between flat and non-flat cosmologies.  If no
+        modifications are requested, then a reference to this object is
+        returned.
+
+        Returns
+        -------
+        newcosmo : `~astropy.cosmology.Cosmology` subclass instance
+            A new instance of this class with updated parameters as specified.
+
+        Examples
+        --------
+        To make a copy of the ``Planck13`` cosmology with a different matter
+        density (``Om0``):
+        and a new name:
+
+        >>> from astropy.cosmology import Planck13
+        >>> newcosmo = Planck13.clone(name="Modified Planck 2013", Om0=0.35)
+        """
+        # Quick return check, taking advantage of the
+        # immutability of cosmological objects
+        if len(kwargs) == 0:
+            return self
 
-class FLRW(Cosmology, metaclass=ABCMeta):
+        # Mix kwargs into initial arguments, preferring the former.
+        full_kwargs = {**self._init_arguments, **kwargs}
+        # Create BoundArgument to handle args versus kwargs.
+        # This also handles all errors from mismatched arguments
+        try:
+            ba = self._init_signature.bind_partial(**full_kwargs)
+        except TypeError as e:
+            # for backward compatibility, map TypeError to AttributeError
+            warnings.warn("Starting in Astropy 5.0, passing an unrecognized "
```

Comment:
Good, yes - very very minor nitpick 😅 
```suggestion
            warnings.warn("Starting in Astropy v5.0, passing an unrecognized "
```

Reply:
Great! I'll commit this and then squash.